File: frontend/src/api/audit_log.ts
/**
 * api/audit_log.ts — 稽核日誌相關 API
 *
 * 稽核日誌由後端自動記錄訂閱的新增、修改、刪除操作，前端只有查詢功能。
 * 必須傳入日期範圍，避免一次拉取全部資料。
 */
import { api } from './client'
import type { ApiResponse, AuditLogEntry } from '@/types/api'

/**
 * 取得指定日期範圍內的稽核日誌。
 * @param fromDate 起始日期（YYYY-MM-DD）
 * @param toDate 結束日期（YYYY-MM-DD）
 */
export async function listAuditLog(
  fromDate: string,
  toDate: string,
): Promise<AuditLogEntry[]> {
  const { data } = await api.get<ApiResponse<AuditLogEntry[]>>('/api/v1/audit-log', {
    params: { from_date: fromDate, to_date: toDate },
  })
  if (!data.success || !data.data) throw new Error(data.message)
  return data.data
}

File: backend/src/api/v1/routers/audit_log.py
from datetime import date

from application.use_cases.list_audit_log import ListAuditLogUseCase
from domain.entities.audit_entry import AuditEntry
from fastapi import APIRouter, Depends, Query
from infrastructure.database.repositories.audit_log_repository import SqlAuditLogRepository
from infrastructure.database.session import get_db
from sqlalchemy.ext.asyncio import AsyncSession

from api.dependencies import require_admin
from api.v1.schemas.audit_log import AuditLogChangeItem, AuditLogResponse
from api.v1.schemas.base import ApiResponse

router = APIRouter(prefix="/api/v1/audit-log", tags=["audit-log"])


def _to_response(entry: AuditEntry) -> AuditLogResponse:
    raw_changes = entry.details.get("changes")
    changes = [AuditLogChangeItem(**c) for c in raw_changes] if raw_changes else None
    return AuditLogResponse(
        id=entry.id,
        action=entry.action,
        resource_id=entry.resource_id,
        user_email=entry.details.get("user_email"),
        service_name=entry.details.get("service_name"),
        changes=changes,
        created_at=entry.created_at,
    )


@router.get("", response_model=ApiResponse[list[AuditLogResponse]])
async def list_audit_log(
    from_date: date = Query(...),
    to_date: date = Query(...),
    _=Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> ApiResponse[list[AuditLogResponse]]:
    repo = SqlAuditLogRepository(db)
    use_case = ListAuditLogUseCase(repo)
    entries = await use_case.execute(from_date=from_date, to_date=to_date)
    return ApiResponse.ok(data=[_to_response(e) for e in entries])
